airbyte-integrations/connectors/source-adaptive/source_adaptive/adaptive/base.py

```diff
@@ -0,0 +1,136 @@
+import requests
+import json
+
+from abc import ABC, abstractmethod
+from airbyte_cdk.logger import AirbyteLogger
+from time import sleep
+from requests.models import Response
+
+from airbyte_cdk.models import (
+    AirbyteMessage,
+)
+
+import csv
+import xmltodict
+from typing import Generator
+from functools import lru_cache, wraps
+from datetime import datetime, timedelta
+
+
+def get_config_as_dict(config: json) -> dict:
+    """
+    Transform config into a simple dict that can be accessed with ease
+    """
+    return json.loads(json.dumps(config))
+
+
+class Adaptive(ABC):
+    """
+    An abstract class to handle all communication with the adaptive,in addition
+    all methods that are not specifically used by one method it is advices to
+    be placed here so any future method can reuse the method.
+    """
+    def __init__(self, logger: AirbyteLogger, config: json):
+        self.logger = logger
+        self.config = get_config_as_dict(config=config) # save config in better format
+
+    def perform_request(self):
+        """
+        A Generic method to perform the request, please do not override this
+        in theory you should be able to perform different kind of request
+        depending on what you want to know. However, the reponse of adaptive
+        is of type xml, which means any kind of operation is very difficult
+        without parsing the response as a whole xml structure. Several optimization
+        could be implemented but reading it as text and parsing the response
+        as stream but this will make the code unmantainable. So for adaptive
+        it is better to be slow since the load data is not so big.
+        """
+        url = "https://api.adaptiveinsights.com/api/v32"
+        headers = {"Content-Type": "application/xml"}
+
+        # have payload ready
+        payload = self.construct_payload()
+
+        response = None
+        counter = 0
+        while response is None:
```

Comment:
Very Nice logic for Networking Error Handling. I would examine a smaller counter value, but it's not that important.

Reply:
necessity is the mother of all inventions. many times I got the error
```
The error "ConnectionError: Max retries exceeded with url" occurs for multiple reasons: An incorrect or incomplete URL has been passed to the request. get() method.
```
there are times where the request might be performed 10 times before it actually succeeds. Dont know why really, this took a lot of researching and I cannot come up with a plausible explanation

Reply:
But yeah, gathering all requests in a central place, and simply calling it from anywhere needed is a good practice which helps the error handling, because networking is always a pain